src/timer/windows/SDL_systimer.c

```diff
@@ -146,18 +146,32 @@ SDL_GetPerformanceFrequency(void)
 void
 SDL_Delay(Uint32 ms)
 {
-    /* Sleep() is not publicly available to apps in early versions of WinRT.
-     *
-     * Visual C++ 2013 Update 4 re-introduced Sleep() for Windows 8.1 and
-     * Windows Phone 8.1.
-     *
-     * Use the compiler version to determine availability.
-     *
-     * NOTE #1: _MSC_FULL_VER == 180030723 for Visual C++ 2013 Update 3.
-     * NOTE #2: Visual C++ 2013, when compiling for Windows 8.0 and
-     *    Windows Phone 8.0, uses the Visual C++ 2012 compiler to build
-     *    apps and libraries.
-     */
+    /* CREATE_WAITABLE_TIMER_HIGH_RESOLUTION flag was added in Windows 10 version 1803.
+    * 
+    * Sleep() is not publicly available to apps in early versions of WinRT.
+    *
+    * Visual C++ 2013 Update 4 re-introduced Sleep() for Windows 8.1 and
+    * Windows Phone 8.1.
+    *
+    * Use the compiler version to determine availability.
+    *
+    * NOTE #1: _MSC_FULL_VER == 180030723 for Visual C++ 2013 Update 3.
+    * NOTE #2: Visual C++ 2013, when compiling for Windows 8.0 and
+    *    Windows Phone 8.0, uses the Visual C++ 2012 compiler to build
+    *    apps and libraries.
+    */
+#ifdef CREATE_WAITABLE_TIMER_HIGH_RESOLUTION 
+    HANDLE timer = CreateWaitableTimerExW(NULL, NULL, CREATE_WAITABLE_TIMER_HIGH_RESOLUTION, TIMER_ALL_ACCESS);
```

Comment:
We should bypass this logic and call directly to `Sleep()` if `ms == 0`, since `Sleep(0)` has special timeslice yielding behavior on Windows that I don't know if waitable timers have. Even if they do have the same behavior, having to 4 syscalls (including creating and destroying an object) to yield isn't ideal.

I think we should also call through to `Sleep()` directly if `SDL_HINT_TIMER_RESOLUTION` is set to 1. No sense in using a waitable timer if the timer resolution is already set such that `Sleep(1)` will work as expected.

Reply:
I've been following along with this thread with interest in using this to help maintain very tight timing loops (480, 960 Hz) in cases where we can't rely on newer gpu driver sync functionality, or where the loop is on another thread which doesn't involve display.

I'm saying this without having tried `CREATE_WAITABLE_TIMER_HIGH_RESOLUTION` yet, but `Sleep(1)` on windows is quite inaccurate (always in the upwards direction, generally 1-2ms sleep) to the point it may still be useful in `Sleep(1)` (and potentially higher) scenarios.

Would be handy to have a way of always using the high resolution version if preferred, rather than have a cutoff `>0` where it stops working.

The sleep-0-yield case sounds correct, though.

Reply:
I didn't really thought about the Sleep(0) case. So I benchmarked both methods, and I could observed that they randomly vary between 10 µs and 500 µs, which seems to indicate that they both have the same behavior and try to sleep for the remainder of the current time-slice (a time slice is around 500 µs).

> I think we should also call through to Sleep() directly if SDL_HINT_TIMER_RESOLUTION is set to 1. No sense in using a waitable timer if the timer resolution is already set such that Sleep(1) will work as expected.

I disagree, and that's the whole point of this PR, timers with the CREATE_WAITABLE_TIMER_HIGH_RESOLUTION are more accurate. From [this article](https://croakingkero.com/tutorials/win32_timing/#Addendum), here's the real sleep time for each method when we ask for a 1 ms sleep  :

timeBeginPeriod (the current SDL behavior with SDL_HINT_TIMER_RESOLUTION set to 1) :
Avg: 1887 ms
Min: 1010 ms
Max: 2689 ms

Timers with CREATE_WAITABLE_TIMER_HIGH_RESOLUTION :
Avg: 701 ms
Min: 504 ms
Max: 1405 ms

Reply:
> I disagree, and that's the whole point of this PR, timers with the CREATE_WAITABLE_TIMER_HIGH_RESOLUTION are more accurate. From [this article](https://croakingkero.com/tutorials/win32_timing/#Addendum), here's the real sleep time for each method when we ask for a 1 ms sleep :
> 
> timeBeginPeriod (the current SDL behavior with SDL_HINT_TIMER_RESOLUTION set to 1) :
> Avg: 1887 ms
> Min: 1010 ms
> Max: 2689 ms
> 
> Timers with CREATE_WAITABLE_TIMER_HIGH_RESOLUTION :
> Avg: 701 ms
> Min: 504 ms
> Max: 1405 ms

[The `SDL_Delay()` docs](https://wiki.libsdl.org/SDL_Delay) say "It waits **at least the specified time**, but possibly longer due to OS scheduling." (emphasis mine). Technically, that means the high resolution timer implementation violates the API contract for `SDL_Delay()` by returning too early, though whether that matters in practice is unclear.

Since we only have millisecond precision in the `SDL_Delay()` parameter, we can't really distinguish "I would like to wait for the shortest possible non-zero period" from "I must wait at least 1 millisecond".  If there's a good use case for shorter than 1 ms waits, maybe we could introduce a new API in SDL3 for that.

All that said, I don't feel strongly that `Sleep()` should be used for the 1 ms wait case. I'm fine with a solution that uses the waitable timer for that (as long as the API consideration above is deemed not to be a blocker).

Reply:
Sorry, what I wrote is wrong. In the benchmarks I quoted, Sleep is called with 1 ms, but the timer allow for greater accuracy and is set to wait for 500 µs, so it does wait for at least the specified time.

As timers can bet set below 1ms, I agree that it could be useful to add a new sleep function that accept the wait time in µs.